consensus/istanbul/backend/announce.go

```diff
@@ -302,83 +390,92 @@ func (sb *Backend) handleIstAnnounce(payload []byte) error {
+	// Check if the sender is within the registered/elected valset
+	regAndActiveVals, err := sb.retrieveActiveAndRegisteredValidators()
 	if err != nil {
-		logger.Warn("Error in decoding received Istanbul Announce message content", "err", err, "IstanbulMsg", msg.String())
 		return err
 	}
 
-	logger = logger.New("msgAddress", msg.Address, "msg_timestamp", announceData.Timestamp)
-
-	if currentEntryTimestamp, err := sb.valEnodeTable.GetTimestampFromAddress(msg.Address); err == nil {
-		if announceData.Timestamp < currentEntryTimestamp {
-			logger.Trace("Received an old announce message", "currentEntryTimestamp", currentEntryTimestamp)
-			return errOldAnnounceMessage
-		}
-	} else if err != leveldb.ErrNotFound {
-		logger.Warn("Error when retrieving timestamp for entry in the ValEnodeTable", "err", err)
+	if !regAndActiveVals[msg.Address] {
+		logger.Debug("Received a message from a non registered/elected validator. Ignoring it.", "sender", msg.Address)
+		return errUnauthorizedAnnounceMessage
 	}
 
-	// If the message is not within the registered validator set, then ignore it
-	regAndActiveVals, err := sb.retrieveActiveAndRegisteredValidators()
+	var announcePayload valEncryptedEnodes
+	err = rlp.DecodeBytes(msg.Msg, &announcePayload)
 	if err != nil {
+		logger.Warn("Error in decoding received Istanbul Announce message content", "err", err, "IstanbulMsg", msg.String())
 		return err
 	}
 
-	if !regAndActiveVals[msg.Address] {
-		logger.Debug("Received an IstanbulAnnounce message from a non registered validator. Ignoring it.", "AnnounceMsg", msg.String(), "err", err)
-		return errUnauthorizedAnnounceMessage
+	logger = logger.New("msgAddress", msg.Address, "msg_timestamp", announcePayload.Timestamp)
+
+	// Ignore the message if it's older than the one that this node currently has persisted and return from this function
+	sb.cachedAnnounceMsgsMu.Lock()
+	defer sb.cachedAnnounceMsgsMu.Unlock()
+	if cachedAnnounceMsgEntry, ok := sb.cachedAnnounceMsgs[msg.Address]; ok && cachedAnnounceMsgEntry.MsgTimestamp >= announcePayload.Timestamp {
+		logger.Trace("Received announce message that has older or same version than the one cached", "cached_timestamp", cachedAnnounceMsgEntry.MsgTimestamp)
+		return errOldAnnounceMessage
 	}
 
-	var node *enode.Node
-	var destAddresses = make([]string, 0, len(announceData.AnnounceRecords))
-	var processedAddresses = make(map[common.Address]bool)
+	// If this is a registered or elected validator, then process the announce message
+	var destAddresses = make([]string, 0, len(announcePayload.EncryptedEnodes))
 	var msgHasDupsOrIrrelevantEntries bool = false
-	for _, announceRecord := range announceData.AnnounceRecords {
-		// Don't process duplicate entries or entries that are not in the regAndActive valset
-		if !regAndActiveVals[announceRecord.DestAddress] || processedAddresses[announceRecord.DestAddress] {
-			msgHasDupsOrIrrelevantEntries = true
-			continue
+	if sb.coreStarted && regAndActiveVals[sb.Address()] {
```

Comment:
Does `sb.coreStarted` need to be accessed under a mutex? I noticed that was done elsewhere but it unclear to me why.
https://github.com/celo-org/celo-blockchain/blob/417be0b1925ad71ea06c70fff511bc0b68f3acd0/consensus/istanbul/backend/announce.go#L90-L105

Reply:
It's not really needed to be protected within a read lock.  Will remove it.